File: frontend/lib/features/upload/domain/stl_file.dart
class STLFile {
  final String id;
  final String originalFilename;
  final int fileSizeBytes;
  final String status; // 'uploaded' | 'analyzing' | 'ready' | 'error'
  final DateTime createdAt;
  final String? glbUrl; // URL vers l'endpoint GET /stl/{id}/glb
  // geometry — nullable jusqu'à status = ready
  final double? volumeCm3;
  final double? surfaceAreaCm2;
  final double? bboxXMm;
  final double? bboxYMm;
  final double? bboxZMm;
  final int? triangleCount;
  final String? hasOverhangs; // 'yes' | 'no' | 'unknown'
  final String? hasThinWalls; // 'yes' | 'no' | 'unknown'


   // Advanced geometry features (Sprint 2B) ──
  final double? overhangRatio;       // 0.0–1.0
  final double? maxOverhangAngle;    // degrees
  final double? minWallThicknessMm;
  final double? avgWallThicknessMm;
  final double? complexityIndex;     // surface_area / volume
  final double? aspectRatio;         // max(bbox) / min(bbox)
  final bool? isWatertight;
  final int? shellCount;
  final double? comOffsetRatio;
  final double? flatBaseAreaMm2;


  STLFile({
    required this.id,
    required this.originalFilename,
    required this.fileSizeBytes,
    required this.status,
    required this.createdAt,
    this.glbUrl,
    this.volumeCm3,
    this.surfaceAreaCm2,
    this.bboxXMm,
    this.bboxYMm,
    this.bboxZMm,
    this.triangleCount,
    this.hasOverhangs,
    this.hasThinWalls,
    this.overhangRatio,
    this.maxOverhangAngle,
    this.minWallThicknessMm,
    this.avgWallThicknessMm,
    this.complexityIndex,
    this.aspectRatio,
    this.isWatertight,
    this.shellCount,
    this.comOffsetRatio,
    this.flatBaseAreaMm2,

  });

  static String? _normalizeFlag(dynamic value) {
    if (value == null) return null;
    if (value is bool) return value ? 'yes' : 'no';

    final lower = value.toString().toLowerCase();
    if (lower == 'yes' || lower == 'no' || lower == 'unknown') {
      return lower;
    }
    return null;
  }

  factory STLFile.fromJson(Map<String, dynamic> json) => STLFile(
    id: json['id'].toString(),
    originalFilename: (json['original_filename'] ?? '').toString(),
    fileSizeBytes: (json['file_size_bytes'] as num).toInt(),
    status: (json['status'] ?? '').toString(),
    createdAt: DateTime.parse(json['created_at']),
    glbUrl: json['glb_url']?.toString(),
    volumeCm3: json['volume_cm3']?.toDouble(),
    surfaceAreaCm2: json['surface_area_cm2']?.toDouble(),
    bboxXMm: json['bbox_x_mm']?.toDouble(),
    bboxYMm: json['bbox_y_mm']?.toDouble(),
    bboxZMm: json['bbox_z_mm']?.toDouble(),
    triangleCount: (json['triangle_count'] as num?)?.toInt(),
    hasOverhangs: _normalizeFlag(json['has_overhangs']),
    hasThinWalls: _normalizeFlag(json['has_thin_walls']),
    // advanced
    overhangRatio: json['overhang_ratio']?.toDouble(),
    maxOverhangAngle: json['max_overhang_angle']?.toDouble(),
    minWallThicknessMm: json['min_wall_thickness_mm']?.toDouble(),
    avgWallThicknessMm: json['avg_wall_thickness_mm']?.toDouble(),
    complexityIndex: json['complexity_index']?.toDouble(),
    aspectRatio: json['aspect_ratio']?.toDouble(),
    isWatertight: json['is_watertight'] as bool?,
    shellCount: (json['shell_count'] as num?)?.toInt(),
    comOffsetRatio: json['com_offset_ratio']?.toDouble(),
    flatBaseAreaMm2: json['flat_base_area_mm2']?.toDouble(),
  );

  String get formattedSize {
    if (fileSizeBytes < 1024 * 1024) {
      return '${(fileSizeBytes / 1024).toStringAsFixed(1)} KB';
    }
    return '${(fileSizeBytes / (1024 * 1024)).toStringAsFixed(1)} MB';
  }

  String get fileExtension => originalFilename.split('.').last.toUpperCase();

  bool get isReady => status == 'ready';
  bool get isAnalyzing => status == 'analyzing';
  bool get isError => status == 'error';

}

File: backend/app/models/stl_file.py
from sqlalchemy import text
import uuid
from datetime import datetime, timezone
from sqlalchemy import Column, String, Integer, Float, DateTime, ForeignKey, Boolean, JSON
from sqlalchemy.dialects.postgresql import UUID
from app.core.database import Base


class STLFile(Base):
    __tablename__ = "stl_files"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    original_filename = Column(String, nullable=False)
    stored_filename = Column(String, nullable=False, unique=True)
    file_size_bytes = Column(Integer, nullable=False)

    # Populated by geometry extraction — all nullable at upload time
    volume_cm3 = Column(Float, nullable=True)
    surface_area_cm2 = Column(Float, nullable=True)
    bbox_x_mm = Column(Float, nullable=True)
    bbox_y_mm = Column(Float, nullable=True)
    bbox_z_mm = Column(Float, nullable=True)
    triangle_count = Column(Integer, nullable=True)
    has_overhangs = Column(Boolean, nullable=True)  # None=unknown, True=yes, False=no
    has_thin_walls = Column(Boolean, nullable=True)

    # NEW — Geometry features (Sprint 2B)
    overhang_ratio = Column(Float, nullable=True)
    max_overhang_angle = Column(Float, nullable=True)
    min_wall_thickness_mm = Column(Float, nullable=True)
    avg_wall_thickness_mm = Column(Float, nullable=True)
    complexity_index = Column(Float, nullable=True)
    aspect_ratio = Column(Float, nullable=True)
    is_watertight = Column(Boolean, nullable=True)
    shell_count = Column(Integer, nullable=True)
    com_offset_ratio = Column(Float, nullable=True)
    flat_base_area_mm2 = Column(Float, nullable=True)
    face_normal_histogram = Column(JSON, nullable=True)

    # NEW — Orientation results (Sprint 2B)
    best_orientation_1 = Column(JSON, nullable=True)
    best_orientation_2 = Column(JSON, nullable=True)
    best_orientation_3 = Column(JSON, nullable=True)
    best_orientation_score = Column(Float, nullable=True)

    # Status progression: 'uploaded' → 'analyzing' → 'ready' | 'error'
    status = Column(String, nullable=False, default="uploaded")

    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )
    updated_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        server_default=text("now()"),
    )

    # Null until conversion has completed successfully.
    glb_filename = Column(String, nullable=True)
